library/psa_crypto.c

```diff
@@ -93,10 +93,25 @@ static int key_type_is_raw_bytes(psa_key_type_t type)
 #define RNG_INITIALIZED 1
 #define RNG_SEEDED 2
 
+typedef enum {
```

Comment:
Can you please explain to me the deadlock we are preventing by the changes in this commit?

Reply:
Thanks for explaining it off-line: can't hold the global mutex during the RNG initialisation, because that might call into MD or cipher, which in turn would try to acquire the mutex to check the driver initialisation state.

Reply:
The RNG initialisation can end up calling into md, which can end up calling psa_can_do_hash(), which (being an external function) has to lock the global data mutex in order to read the driver initialisation state.

This meant that we had to have more than one mutex, but also as a knock on meant that we would have to release the global data mutex in order to initialise the RNG.

My initial implementation was to do the whole of init atomically - this was the obvious simple implementation: lock, check the flag, then init (if required) then unlock. Now that we had to unlock I had to split the init into atomic  'chunks'. I couldn't split the init state into more stages via an enum, because I couldn't prevent the race condition in deciding which thread did the next step of initialisation. The only solution was to split into submodules, each with an initialisation flag, and each doing the "lock / check flag / init if necessary" step. This obviously meant that atomic init chunks could be done in parallel, which is why the additional check of the drivers being initialised in the RNG init, but other than that , testing has not shown any further issues.

Reply:
I was thinking that another way to resolve this would be to not take the lock in `can_do`:
- It only reads the flag and the flag is written only once during the whole life-cycle (if we consider `mbedtls_psa_crypto_free()` out of scope which I think we should do)
- Even if there is a race condition, there is a linearisation point before the call to the init:
    - if the `can_do` is called from the init, then we know that no other thread is writing it as we hold the global mutex
    - if the `can_do` is called from somewhere else, then it is a parallel call without a preceding init and the linearisation point for it is correct for both before and after the init call and whichever behaviour we end up with is thread safe.

Reply:
(Note: I haven't read the code, only the discussion here.)

> The only solution was to split into submodules, each with an initialisation flag, and each doing the "lock / check flag / init if necessary" step.

This is something we will need in the future: Mbed TLS 4.0 will support staged initialization, where the application requests initialization of e.g. just the keystore, or just the RNG. It doesn't have to be thread-safe (this is essentially only useful for bootloaders that can't assume that the system is ready for everything, and those are typically not multithreaded), but it's nicer if it is.